Add clone and single vm concurrent tests

brunotm opened the pull request:
Fixes #772

msterin commented on vmdk_plugin/sanity_test.go:
```diff
@@ -240,7 +242,7 @@ func TestSanity(t *testing.T) {
 		},
 	}
-	// Create/delete routine
+	// Create/delete goroutine
```
So the problem here - which I observed on my testbed-  is that if there is  a client  in clients more than once, the code below (go func) fails because ESX rejects reconfigure request on the same VM when another one is in flight. 

and our current Makefiles allow for VM1 and  VM2 (test clients) to be the same - in fact this is how single VM testing is done. 

I'd suggest  checking for this  and warn then skip duplicate clients. Yes, it will skip the parallel test but  will unblock single VM test.

brunotm replied:
I wasn't aware of that possibility, or the fact that esx would error out. But independently of skipping the tests for that specific scenario we should serialize reconfig tasks per vm in vmdk_ops.

msterin commented on vmdk_plugin/sanity_test.go:
```diff
@@ -260,4 +262,54 @@ func TestSanity(t *testing.T) {
+	for idx, elem := range clients {
+		go func(idx int, c *client.Client) {
+			for i := 0; i < parallelClones; i++ {
+				volName := masterVolName + "-clone" + strconv.Itoa(idx) + strconv.Itoa(i)
```
nit: formatting with fmt. is easier to read IMO.

brunotm replied:
Ok

msterin commented on vmdk_plugin/sanity_test.go:
```diff
@@ -260,4 +262,54 @@ func TestSanity(t *testing.T) {
+	}
+
+	// Read the results from the channel
+	for i := 0; i < len(clients)*parallelClones*2; i++ {
```
nit: do we need spaces around * ?

brunotm replied:
This is done by gofmt.

msterin replied:
ignore my comment about * then :-)

msterin commented on vmdk_plugin/sanity_test.go:
```diff
@@ -41,6 +41,7 @@ const (
 	defaultTestLogPath = "/tmp/test-docker-volume-vsphere.log"
 	// Number of volumes per client for parallel tests
 	parallelVolumes = 9
+	parallelClones  = 3
```
nit: both these constants seem prime candidates for test arguments

brunotm replied:
Yes :)

brunotm commented:
@msterin 
I was thinking about also pushing in this PR single vm concurrent tests and address your comments. The corresponding issue was updated to reflect the single vm tests. Unless you think they should go in a different one. :)

msterin commented:
> @msterin I was thinking about also pushing in this PR single vm concurrent tests and address your comments. The corresponding issue was updated to reflect the single vm tests. Unless you think they should go in a different one. :)

IMO either way is fine as long as there are 2 separate commits

brunotm commented:
@msterin 
I have included the commit from #810 here, otherwise tests would fail.
Hopefully this PR together with #810, will be the last edge for the concurrency case.

brunotm commented:
@msterin 
Updated the patch to address the last comment and reverted the commit included from #810.
The CI run might fail without #810.

From my part is good to go.

Thanks.

msterin commented:
> The CI run might fail without #810.

And it does fail: 
```
	sanity_test.go:189: Successfully connected to tcp://192.168.31.206:2375
	sanity_test.go:313: Same docker host concurrent create/delete test failed, err: Error response from daemon: create volTestP-same20: Post http://%2Frun%2Fdocker%2Fplugins%2Fvmdk.sock/VolumeDriver.Create: http: ContentLength=49 with Body length 0
FAIL
```
Until #810 is in, this one has to wait to avoid breaking the master

brunotm commented:
@msterin 
Definitely! :)

brunotm commented:
@msterin 
The concurrent tests are good to go after #810.
